irohad/network/impl/block_loader_service.cpp

```diff
@@ -56,21 +56,45 @@ grpc::Status BlockLoaderService::retrieveBlock(
                         "Bad hash provided");
   }
 
-  // required block must be in the cache
+  // try to fetch block from the consensus cache
   auto block = consensus_result_cache_->get();
-  if (not block) {
-    log_->info("Requested to retrieve a block from an empty cache");
-    return grpc::Status(grpc::StatusCode::NOT_FOUND, "Cache is empty");
-  }
-  if (block->hash() != hash) {
+  if (block) {
+    if (block->hash() == hash) {
+      auto transport_block =
+          std::static_pointer_cast<shared_model::proto::Block>(block)
+              ->getTransport();
+      response->CopyFrom(transport_block);
+      return grpc::Status::OK;
+    }
+    log_->info(
+        "Requested to retrieve a block, but cache contains another block: "
+        "requested {}, in cache {}",
+        hash.hex(),
+        block->hash().hex());
+  } else {
     log_->info(
-        "Requested to retrieve a block with hash other than the one in cache");
+        "Tried to retrieve a block from an empty cache: requested block hash "
+        "{}",
+        hash.hex());
+  }
+
+  // cache missed: notify and try to fetch the block from block storage itself
+  auto blocks = block_query_factory_->createBlockQuery() |
+      // TODO [IR-1757] Akvinikym 12.10.18: use block height to get one block
+      // instead of the whole chain
+      [](const auto &block_query) { return block_query->getBlocksFrom(1); };
```

Comment:
`getBlocksFrom` should be wrapped in `optional`, otherwise a default constructed lambda return type will be returned, which is an empty vector, due to `operator|` implementation. While this code handles it correctly, it is still not obvious to the reader.